variables.tf

```diff
@@ -26,6 +26,29 @@ variable "acl" {
   description = "The canned ACL to apply, defaults to `private`."
 }
 
+variable "use_acl" {
+  description = "Whether to use canned ACL (true) or access_control_policy (false)"
+  type        = bool
+  default     = true
+}
+
+variable "access_control_policy" {
```

Comment:
you do a lot of validation in the main.tf, could you move that outside to the variable and use validation for this? 

besides, could we simplify this variable? e.g. 

```
owner = object({
      id = string
    })
```

```
owner_id = string
```

Reply:
variable "access_control_policy" {
  type = object({
    owner_id = string
    grants = list(object({
      grantee = object({
        type       = string # Allowed values: "CanonicalUser", "Group", "AmazonCustomerByEmail"
        identifier = string # Maps to id, uri, or email_address based on the grantee type
      })
      permission = string
    }))
  })

  default = null # Making it optional
}
This is what i mean with the variable, owner and id do not need to be nested. Then if you add validation for the type in grantee to check for "CanonicalUser", "Group", "AmazonCustomerByEmail" then it looks great to me